ansible/roles/images/tasks/main.yaml

```diff
@@ -4,15 +4,30 @@
     name: containerd
     state: restarted
 
-- name: pull control plane images
-  shell: "crictl pull {{ item }}"
-  loop: "{{ control_plane_images }}"
+- name: determine kubernetes images
+  shell: kubeadm config images list --image-repository {{ k8s_image_registry }}
+  register: kubernetes_kubeadm_images
 
-- name: pull extra images
-  shell: "crictl pull {{ item }}"
-  loop: "{{ extra_images }}"
+- name: determine containerd pause image
+  shell: >
+    containerd config default |
+    sed -n 's/^.*sandbox_image\s*=\s"\(\S*\)".*$/\1/p'
+  register: containerd_sandbox_image
 
-- name: pull aws images
-  shell: "crictl pull {{ item }}"
-  loop: "{{ aws_images }}"
+- name: set images fact
+  set_fact:
+    pull_images: "{{
+      kubernetes_kubeadm_images.stdout_lines +
+      containerd_sandbox_image.stdout_lines +
+      control_plane_images +
+      extra_images
+    }}"
+
+- name: set aws images
+  set_fact:
+    pull_images: "{{ pull_images + aws_images }}"
   when: packer_builder_type.startswith('amazon')
+
+- name: pull images
+  shell: "crictl pull {{ item }}"
+  loop: "{{ pull_images | replace_images(image_replacements | default({})) }}"
```

Comment:
Can you please explain what `replace_images` is doing.

Reply:
It's replacing `  "k8s.gcr.io/etcd:3.4.13-0"` with `"{{ k8s_image_registry }}/etcd:v3.4.13_fips.0"` from the map, `image_replacments` (see `overrides/fips.yaml` below).

Reply:
Correct, I added a small note for the time being, but I'm going to see if we can get rid of the `image_replacements` all together.

Reply:
I was able to get rid of this completely and move it into the playbook. Now the `build tag` from `kubeadm` drives the whole process of setting the `fips.0` (et. al)  tag.